File: Source/Client/Forms/Editors/FrmEditor_Pet.vb
Imports System.IO
Imports Core
Friend Class frmEditor_Pet

#Region "Basics"

    Private Sub frmEditor_Pet_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        nudSprite.Maximum = NumCharacters
        nudRange.Maximum = 50
        nudLevel.Maximum = MAX_LEVEL
        nudMaxLevel.Maximum = MAX_LEVEL

        lstIndex.Items.Clear()

        ' Add the names
        For i = 1 To MAX_PETS
            lstIndex.Items.Add(i & ": " & Trim$(Pet(i).Name))
        Next

        cmbEvolve.Items.Clear()
        cmbEvolve.Items.Add("None")
        ' Add the names
        For i = 1 To MAX_PETS
            cmbEvolve.Items.Add(i & ": " & Trim$(Pet(i).Name))
        Next
    End Sub

    Private Sub LstIndex_Click(sender As Object, e As EventArgs) Handles lstIndex.Click
        PetEditorInit()
    End Sub

    Private Sub TxtName_TextChanged(sender As Object, e As EventArgs) Handles txtName.TextChanged
        Dim tmpindex As Integer

        tmpindex = lstIndex.SelectedIndex
        Pet(EditorIndex).Name = Trim$(txtName.Text)
        lstIndex.Items.RemoveAt(EditorIndex - 1)
        lstIndex.Items.Insert(EditorIndex - 1, EditorIndex & ": " & Pet(EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex
    End Sub

    Private Sub NudSprite_Click(sender As Object, e As EventArgs) Handles nudSprite.Click
        Pet(EditorIndex).Sprite = nudSprite.Value

        EditorPet_DrawPet()
    End Sub

    Friend Sub EditorPet_DrawPet()
        Dim petnum As Integer

        petnum = nudSprite.Value

        If petnum <= 0 OrElse petnum > NumCharacters Then
            picSprite.BackgroundImage = Nothing
            Exit Sub
        End If

        If File.Exists(Paths.Graphics & "Characters\" & petnum & GfxExt) Then
            picSprite.Width = Drawing.Image.FromFile(Paths.Graphics & "characters\" & petnum & GfxExt).Width/4
            picSprite.Height = Drawing.Image.FromFile(Paths.Graphics & "characters\" & petnum & GfxExt).Height/4
            picSprite.BackgroundImage = Image.FromFile(Paths.Graphics & "Characters\" & petnum & GfxExt)
        Else
            picSprite.BackgroundImage = Nothing
        End If

    End Sub

    Private Sub NudRange_ValueChanged(sender As Object, e As EventArgs) Handles nudRange.Click
        Pet(EditorIndex).Range = nudRange.Value
    End Sub

    Private Sub BtnSave_Click(sender As Object, e As EventArgs) Handles btnSave.Click
        PetEditorOk()
        Dispose()
    End Sub

    Private Sub BtnCancel_Click(sender As Object, e As EventArgs) Handles btnCancel.Click
        PetEditorCancel()
        Dispose()
End Sub





#End Region
#Region "Stats"

    Private Sub OptCustomStats_CheckedChanged(sender As Object, e As EventArgs) Handles optCustomStats.CheckedChanged
        If optCustomStats.Checked = True Then
            pnlCustomStats.Visible = True
            Pet(EditorIndex).StatType = 1
        Else
            pnlCustomStats.Visible = False
            Pet(EditorIndex).StatType = 0
        End If
    End Sub

    Private Sub OptAdoptStats_CheckedChanged(sender As Object, e As EventArgs) Handles optAdoptStats.CheckedChanged
        If optAdoptStats.Checked = True Then
            pnlCustomStats.Visible = False
            Pet(EditorIndex).StatType = 0
        Else
            pnlCustomStats.Visible = True
            Pet(EditorIndex).StatType = 1
        End If
    End Sub

    Private Sub NudStrength_ValueChanged(sender As Object, e As EventArgs) Handles nudStrength.ValueChanged
        Pet(EditorIndex).Stat(StatType.Strength) = nudStrength.Value
    End Sub

    Private Sub NudVitality_ValueChanged(sender As Object, e As EventArgs) Handles nudVitality.ValueChanged
        Pet(EditorIndex).Stat(StatType.Vitality) = nudVitality.Value
    End Sub

    Private Sub NudLuck_ValueChanged(sender As Object, e As EventArgs) Handles nudLuck.ValueChanged
        Pet(EditorIndex).Stat(StatType.Luck) = nudLuck.Value
    End Sub

    Private Sub NudIntelligence_ValueChanged(sender As Object, e As EventArgs) Handles nudIntelligence.ValueChanged
        Pet(EditorIndex).Stat(StatType.Intelligence) = nudIntelligence.Value
    End Sub

    Private Sub NudSpirit_ValueChanged(sender As Object, e As EventArgs) Handles nudSpirit.ValueChanged
        Pet(EditorIndex).Stat(StatType.Spirit) = nudSpirit.Value
    End Sub

    Private Sub NudLevel_ValueChanged(sender As Object, e As EventArgs) Handles nudLevel.ValueChanged
        Pet(EditorIndex).Level = nudLevel.Value
    End Sub




#End Region
#Region "Leveling"

    Private Sub NudPetExp_ValueChanged(sender As Object, e As EventArgs) Handles nudPetExp.Click
        Pet(EditorIndex).ExpGain = nudPetExp.Value
    End Sub

    Private Sub NudPetPnts_ValueChanged(sender As Object, e As EventArgs) Handles nudPetPnts.Click
        Pet(EditorIndex).LevelPnts = nudPetPnts.Value
    End Sub

    Private Sub NudMaxLevel_ValueChanged(sender As Object, e As EventArgs) Handles nudMaxLevel.Click
        Pet(EditorIndex).MaxLevel = nudMaxLevel.Value
    End Sub

    Private Sub OptLevel_CheckedChanged(sender As Object, e As EventArgs) Handles optLevel.Click
        If optLevel.Checked = True Then
            pnlPetlevel.Visible = True
            Pet(EditorIndex).LevelingType = 1
        End If
    End Sub

    Private Sub OptDoNotLevel_CheckedChanged(sender As Object, e As EventArgs) Handles optDoNotLevel.Click
        If optDoNotLevel.Checked = True Then
            pnlPetlevel.Visible = False
            Pet(EditorIndex).LevelingType = 0
End If
    End Sub




#End Region
#Region "Skills"

    Private Sub CmbSkill1_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbSkill1.SelectedIndexChanged
        Pet(EditorIndex).Skill(1) = cmbSkill1.SelectedIndex
    End Sub

    Private Sub CmbSkill2_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbSkill2.SelectedIndexChanged
        Pet(EditorIndex).Skill(2) = cmbSkill2.SelectedIndex
    End Sub

    Private Sub CmbSkill3_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbSkill3.SelectedIndexChanged
        Pet(EditorIndex).Skill(3) = cmbSkill3.SelectedIndex
    End Sub

    Private Sub CmbSkill4_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbSkill4.SelectedIndexChanged
        Pet(EditorIndex).Skill(4) = cmbSkill4.SelectedIndex
    End Sub




#End Region
#Region "Evolving"

    Private Sub ChkEvolve_CheckedChanged(sender As Object, e As EventArgs) Handles chkEvolve.CheckedChanged
        If chkEvolve.Checked = True Then
            Pet(EditorIndex).Evolvable = 1
        Else
            Pet(EditorIndex).Evolvable = 0
        End If
    End Sub

    Private Sub NudEvolveLvl_ValueChanged(sender As Object, e As EventArgs) Handles nudEvolveLvl.ValueChanged
        Pet(EditorIndex).EvolveLevel = nudEvolveLvl.Value
    End Sub

    Private Sub CmbEvolve_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cmbEvolve.SelectedIndexChanged
        Pet(EditorIndex).EvolveNum = cmbEvolve.SelectedIndex
    End Sub

    Private Sub btnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
        Dim tmpindex As Integer

        ClearPet(EditorIndex)

        tmpindex = lstIndex.SelectedIndex
        lstIndex.Items.RemoveAt(EditorIndex - 1)
        lstIndex.Items.Insert(EditorIndex - 1, EditorIndex & ": " & Pet(EditorIndex).Name)
        lstIndex.SelectedIndex = tmpindex

        PetEditorInit()
    End Sub

    Private Sub frmEditor_Pet_FormClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        PetEditorCancel
    End Sub

#End Region

End Class

File: Source/Client/Forms/FrmEditor_Pet.cs
using System;
using System.IO;
using System.Windows.Forms;
using Core;
using Microsoft.VisualBasic;

namespace Client
{
    internal partial class frmEditor_Pet
    {
        public frmEditor_Pet()
        {
            InitializeComponent();
        }

        #region Basics

        private void frmEditor_Pet_Load(object sender, EventArgs e)
        {
            nudSprite.Maximum = GameState.NumCharacters;
            nudRange.Maximum = 50m;
            nudLevel.Maximum = Constant.MAX_LEVEL;
            nudMaxLevel.Maximum = Constant.MAX_LEVEL;

            lstIndex.Items.Clear();

            // Add the names
            for (int i = 0; i < Constant.MAX_PETS; i++)
                lstIndex.Items.Add(i + 1 + ": " + Core.Type.Pet[i].Name);

            cmbEvolve.Items.Clear();
            // Add the names
            for (int i = 0; i < Constant.MAX_PETS; i++)
                cmbEvolve.Items.Add(i + 1 + ": " + Core.Type.Pet[i].Name);
        }

        protected override void WndProc(ref Message m)
        {
            const int WM_MOUSEACTIVATE = 0x0021;
            const int WM_NCHITTEST = 0x0084;

            if (m.Msg == WM_MOUSEACTIVATE)
            {
                // Immediately activate and process the click.
                m.Result = new IntPtr(1); // MA_ACTIVATE
                return;
            }
            else if (m.Msg == WM_NCHITTEST)
            {
                // Let the window know the mouse is in client area.
                m.Result = new IntPtr(1); // HTCLIENT
                return;
            }

            base.WndProc(ref m);
        }

        private void lstIndex_Click(object sender, EventArgs e)
        {
            Editors.PetEditorInit();
        }

        private void TxtName_TextChanged(object sender, EventArgs e)
        {
            int tmpindex;

            tmpindex = lstIndex.SelectedIndex;
            Core.Type.Pet[GameState.EditorIndex].Name = Strings.Trim(txtName.Text);
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Type.Pet[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;
        }

        private void NudSprite_Click(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Sprite = (int)Math.Round(nudSprite.Value);

            EditorPet_DrawPet();
        }

        internal void EditorPet_DrawPet()
        {
            int petNum;

            petNum = (int)Math.Round(nudSprite.Value);

            if (petNum <= 0 | petNum > GameState.NumCharacters)
            {
                picSprite.BackgroundImage = null;
                return;
            }

            if (File.Exists(System.IO.Path.Combine(Core.Path.Characters, petNum + GameState.GfxExt)))
            {
                picSprite.Width = (int)Math.Round(System.Drawing.Image.FromFile(System.IO.Path.Combine(Core.Path.Characters, petNum + GameState.GfxExt)).Width / 4d);
                picSprite.Height = (int)Math.Round(System.Drawing.Image.FromFile(System.IO.Path.Combine(Core.Path.Characters, petNum + GameState.GfxExt)).Height / 4d);
                picSprite.BackgroundImage = System.Drawing.Image.FromFile(System.IO.Path.Combine(Core.Path.Characters, petNum + GameState.GfxExt));
            }
            else
            {
                picSprite.BackgroundImage = null;
            }

        }

        private void NudRange_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Range = (int)Math.Round(nudRange.Value);
        }

        private void BtnSave_Click(object sender, EventArgs e)
        {
            Editors.PetEditorOK();
            Dispose();
        }

        private void BtnCancel_Click(object sender, EventArgs e)
        {
            Editors.PetEditorCancel();
            Dispose();
        }

        #endregion
        #region Stats

        private void OptCustomStats_CheckedChanged(object sender, EventArgs e)
        {
            if (optCustomStats.Checked == true)
            {
                pnlCustomStats.Visible = true;
                Core.Type.Pet[GameState.EditorIndex].StatType = 1;
            }
            else
            {
                pnlCustomStats.Visible = false;
                Core.Type.Pet[GameState.EditorIndex].StatType = 0;
            }
        }

        private void OptAdoptStats_CheckedChanged(object sender, EventArgs e)
        {
            if (optAdoptStats.Checked == true)
            {
                pnlCustomStats.Visible = false;
                Core.Type.Pet[GameState.EditorIndex].StatType = 0;
            }
            else
            {
                pnlCustomStats.Visible = true;
                Core.Type.Pet[GameState.EditorIndex].StatType = 1;
            }
        }

        private void NudStrength_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Stat[(int)Core.Enum.StatType.Strength] = (byte)Math.Round(nudStrength.Value);
        }

        private void NudVitality_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Stat[(int)Core.Enum.StatType.Vitality] = (byte)Math.Round(nudVitality.Value);
        }

        private void NudLuck_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Stat[(int)Core.Enum.StatType.Luck] = (byte)Math.Round(nudLuck.Value);
        }

        private void NudIntelligence_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Stat[(int)Core.Enum.StatType.Intelligence] = (byte)Math.Round(nudIntelligence.Value);
        }

        private void NudSpirit_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Stat[(int)Core.Enum.StatType.Spirit] = (byte)Math.Round(nudSpirit.Value);
        }

        private void NudLevel_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Level = (byte)Math.Round(nudLevel.Value);
        }




        #endregion
        #region Leveling

        private void NudPetExp_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].ExpGain = (int)Math.Round(nudPetExp.Value);
        }

        private void NudPetPnts_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Points = (byte)Math.Round(nudPetPnts.Value);
        }

        private void NudMaxLevel_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].MaxLevel = (int)Math.Round(nudMaxLevel.Value);
        }

        private void OptLevel_CheckedChanged(object sender, EventArgs e)
        {
            if (optLevel.Checked == true)
            {
                pnlPetlevel.Visible = true;
                Core.Type.Pet[GameState.EditorIndex].LevelingType = 1;
            }
        }

        private void OptDoNotLevel_CheckedChanged(object sender, EventArgs e)
        {
            if (optDoNotLevel.Checked == true)
            {
                pnlPetlevel.Visible = false;
                Core.Type.Pet[GameState.EditorIndex].LevelingType = 0;
            }
        }




        #endregion
        #region Skills

        private void CmbSkill1_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Skill[0] = cmbSkill1.SelectedIndex;
        }

        private void CmbSkill2_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Skill[1] = cmbSkill2.SelectedIndex;
        }

        private void CmbSkill3_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Skill[2] = cmbSkill3.SelectedIndex;
        }

        private void CmbSkill4_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].Skill[3] = cmbSkill4.SelectedIndex;
        }

        #endregion
        #region Evolving

        private void ChkEvolve_CheckedChanged(object sender, EventArgs e)
        {
            if (chkEvolve.Checked == true)
            {
                Core.Type.Pet[GameState.EditorIndex].Evolvable = 1;
            }
            else
            {
                Core.Type.Pet[GameState.EditorIndex].Evolvable = 0;
            }
        }

        private void NudEvolveLvl_ValueChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].EvolveLevel = (int)Math.Round(nudEvolveLvl.Value);
        }

        private void CmbEvolve_SelectedIndexChanged(object sender, EventArgs e)
        {
            Core.Type.Pet[GameState.EditorIndex].EvolveNum = cmbEvolve.SelectedIndex;
        }

        private void btnDelete_Click(object sender, EventArgs e)
        {
            int tmpindex;

            Pet.ClearPet(GameState.EditorIndex);

            tmpindex = lstIndex.SelectedIndex;
            lstIndex.Items.RemoveAt(GameState.EditorIndex);
            lstIndex.Items.Insert(GameState.EditorIndex, GameState.EditorIndex + 1 + ": " + Core.Type.Pet[GameState.EditorIndex].Name);
            lstIndex.SelectedIndex = tmpindex;

            Editors.PetEditorInit();
        }

        private void frmEditor_Pet_FormClosing(object sender, FormClosingEventArgs e)
        {
            Editors.PetEditorCancel();
        }

        #endregion

    }
}